drivers/store/redis/store.go

```diff
@@ -52,263 +85,107 @@ func NewStoreWithOptions(client Client, options limiter.StoreOptions) (limiter.S
+	if ttl > 0 {
+		expiration = now.Add(time.Duration(ttl) * time.Millisecond)
+	}
+	return common.GetContextFromState(now, rate, expiration, count), nil
 }
 
 // Reset returns the limit for given identifier which is set to zero.
 func (store *Store) Reset(ctx context.Context, key string, rate limiter.Rate) (limiter.Context, error) {
 	key = fmt.Sprintf("%s:%s", store.Prefix, key)
-	now := time.Now()
-
-	lctx := limiter.Context{}
-	onWatch := func(rtx *libredis.Tx) error {
-
-		err := store.doResetValue(ctx, rtx, key)
-		if err != nil {
-			return err
-		}
-
-		count := int64(0)
-		expiration := now.Add(rate.Period)
-
-		lctx = common.GetContextFromState(now, rate, expiration, count)
-		return nil
-	}
-
-	err := store.client.Watch(ctx, onWatch, key)
-	if err != nil {
-		err = errors.Wrapf(err, "limiter: cannot reset value for %s", key)
+	if _, err := store.client.Del(ctx, key).Result(); err != nil {
 		return limiter.Context{}, err
 	}
-
-	return lctx, nil
+	count := int64(0)
+	now := time.Now()
+	expiration := now.Add(rate.Period)
+	return common.GetContextFromState(now, rate, expiration, count), nil
 }
 
-// doPeekValue will execute peekValue with a retry mecanism (optimistic locking) until store.MaxRetry is reached.
-func (store *Store) doPeekValue(ctx context.Context, rtx *libredis.Tx, key string) (int64, time.Duration, error) {
-	for i := 0; i < store.MaxRetry; i++ {
-		count, ttl, err := peekValue(ctx, rtx, key)
-		if err == nil {
-			return count, ttl, nil
+// evalSHA eval the redis lua sha and load the script if missing
+func (store *Store) evalSHA(ctx context.Context, sha string, keys []string, args ...interface{}) *libredis.Cmd {
+	cmd := store.client.EvalSha(ctx, sha, keys, args...)
+	if err := cmd.Err(); err != nil {
+		if !isLuaScriptGone(err) {
+			return cmd
 		}
-	}
-	return 0, 0, errors.New("retry limit exceeded")
-}
-
-// peekValue will retrieve the counter and its expiration for given key.
-func peekValue(ctx context.Context, rtx *libredis.Tx, key string) (int64, time.Duration, error) {
-	pipe := rtx.TxPipeline()
-	value := pipe.Get(ctx, key)
-	expire := pipe.PTTL(ctx, key)
-
-	_, err := pipe.Exec(ctx)
-	if err != nil && err != libredis.Nil {
-		return 0, 0, err
-	}
-
-	count, err := value.Int64()
-	if err != nil && err != libredis.Nil {
-		return 0, 0, err
-	}
-
-	ttl, err := expire.Result()
-	if err != nil {
-		return 0, 0, err
-	}
-
-	return count, ttl, nil
-}
-
-// doSetValue will execute setValue with a retry mecanism (optimistic locking) until store.MaxRetry is reached.
-func (store *Store) doSetValue(ctx context.Context, rtx *libredis.Tx,
-	key string, expiration time.Duration) (bool, error) {
-
-	for i := 0; i < store.MaxRetry; i++ {
-		created, err := setValue(ctx, rtx, key, expiration)
-		if err == nil {
-			return created, nil
+		if err := store.preloadLuaScripts(ctx); err != nil {
+			cmd = libredis.NewCmd(ctx)
+			cmd.SetErr(err)
+			return cmd
 		}
+		cmd = store.client.EvalSha(ctx, store.luaPeekSHA, keys)
```

Comment:
Should you retry the given script, instead of peek ?

Reply:
yeah, sorry for my careless